src/clib/pio_darray.c

```diff
@@ -404,6 +521,376 @@ int find_var_fillvalue(file_desc_t *file, int varid, var_desc_t *vdesc)
+    } \
+    else if (iodesc->piotype == FROM_TYPE_ID && av->nc_type == PIO_USHORT) \
+    { \
+        ADIOS_CONVERT_ARRAY(array, arraylen, from_type, unsigned short int, *ierr, buf); \
+    } \
+    else if (iodesc->piotype == FROM_TYPE_ID && av->nc_type == PIO_INT64) \
+    { \
+        ADIOS_CONVERT_ARRAY(array, arraylen, from_type, int64_t, *ierr, buf); \
+    } \
+    else if (iodesc->piotype == FROM_TYPE_ID && av->nc_type == PIO_UINT64) \
+    { \
+        ADIOS_CONVERT_ARRAY(array, arraylen, from_type, uint64_t, *ierr, buf); \
+    } \
+    else if (iodesc->piotype == FROM_TYPE_ID && av->nc_type == PIO_CHAR) \
+    { \
+        ADIOS_CONVERT_ARRAY(array, arraylen, from_type, char, *ierr, buf); \
+    } \
+    else if (iodesc->piotype == FROM_TYPE_ID && av->nc_type == PIO_BYTE) \
+    { \
+        ADIOS_CONVERT_ARRAY(array, arraylen, from_type, char, *ierr, buf); \
+    } \
+    else if (iodesc->piotype == FROM_TYPE_ID && av->nc_type == PIO_UBYTE) \
+    { \
+        ADIOS_CONVERT_ARRAY(array, arraylen, from_type, unsigned char, *ierr, buf); \
+    } \
+}
+
+static void *PIOc_convert_buffer_adios(file_desc_t *file, io_desc_t *iodesc,
+                                       adios_var_desc_t *av, void *array, int arraylen,
+                                       int *ierr)
+{
+    void *buf = array;
+    *ierr = 0;
+
+    ADIOS_CONVERT_FROM(PIO_DOUBLE, double);
+    ADIOS_CONVERT_FROM(PIO_FLOAT, float);
+    ADIOS_CONVERT_FROM(PIO_REAL, float);
+    ADIOS_CONVERT_FROM(PIO_INT, int);
+    ADIOS_CONVERT_FROM(PIO_UINT, unsigned int);
+    ADIOS_CONVERT_FROM(PIO_SHORT, short int);
+    ADIOS_CONVERT_FROM(PIO_USHORT, unsigned short int);
+    ADIOS_CONVERT_FROM(PIO_INT64, int64_t);
+    ADIOS_CONVERT_FROM(PIO_UINT64, uint64_t);
+    ADIOS_CONVERT_FROM(PIO_CHAR, char);
+    ADIOS_CONVERT_FROM(PIO_BYTE, char);
+    ADIOS_CONVERT_FROM(PIO_UBYTE, unsigned char);
+
+    return buf;
+}
+
+#define ADIOS_COPY_ONE(temp_buf, array, var_type) \
+{ \
+    temp_buf = (var_type*)malloc(2 * sizeof(var_type)); \
+    assert(temp_buf != NULL); \
+    memcpy(temp_buf, array, sizeof(var_type)); \
+}
+
+void *PIOc_copy_one_element_adios(void *array, adios_var_desc_t *av)
+{
+    void *temp_buf = NULL;
+    if (av->nc_type == PIO_DOUBLE)
+    {
+        ADIOS_COPY_ONE(temp_buf, array, double);
+    }
+    else if (av->nc_type == PIO_FLOAT || av->nc_type == PIO_REAL)
+    {
+        ADIOS_COPY_ONE(temp_buf, array, float);
+    }
+    else if (av->nc_type == PIO_INT || av->nc_type == PIO_UINT)
+    {
+        ADIOS_COPY_ONE(temp_buf, array, int);
+    }
+    else if (av->nc_type == PIO_SHORT || av->nc_type == PIO_USHORT)
+    {
+        ADIOS_COPY_ONE(temp_buf, array, short int);
+    }
+    else if (av->nc_type == PIO_INT64 || av->nc_type == PIO_UINT64)
+    {
+        ADIOS_COPY_ONE(temp_buf, array, int64_t);
+    }
+    else if (av->nc_type == PIO_CHAR || av->nc_type == PIO_BYTE || av->nc_type == PIO_UBYTE)
+    {
+        ADIOS_COPY_ONE(temp_buf, array, char);
```

Comment:
Using preprocessor macros instead of C functions is not a good idea. The debugger doesn't work and it's hard for other developers to follow. Instead, use C functions.